### DISCUSSION TITLE
Heterogenous graph

### DISCUSSION THREAD

**Discussion by negarmaleki96** [pyg-team/pytorch_geometric#4608] [Q&A]
Hi all,
I am building a heterogeneous graph. I have 3 different node types with 9 additional features within each. Also, I have 3 different edge types with two features within each. I want to implement GNN models on my graph after creating it. My problem is that in the GNN models, they can get only one feature of the graph; for example, in the following code, the model gets x_dict which is common among all the nodes in the heterogeneous graph, but in my case, since I have 9 different features, how should I give it to the model?

class HGT(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels, num_heads, num_layers):
        super().__init__()

        self.lin_dict = torch.nn.ModuleDict()
        for node_type in data.node_types:
            self.lin_dict[node_type] = Linear(-1, hidden_channels)

        self.convs = torch.nn.ModuleList()
        for _ in range(num_layers):
            conv = HGTConv(hidden_channels, hidden_channels, data.metadata(),
                           num_heads, group='sum')
        self.convs.append(conv)

        self.lin = Linear(hidden_channels, out_channels)

    def forward(self, x_dict, edge_index_dict):
        for node_type, x in x_dict.items():
            x_dict[node_type] = self.lin_dict[node_type](x).relu_()

        for conv in self.convs:
            x_dict = conv(x_dict, edge_index_dict)

        return self.lin(x_dict['author'])

model = HGT(hidden_channels=64, out_channels=dataset.num_classes,
            num_heads=2, num_layers=2)

My graph is like this:

HeteroData(
  **post**={
    otype=[147, 1],
    created=[147, 1],
    last_payout=[147, 1],
    payout=[147, 1],
    net_rshares=[147, 1],
    abs_rshares=[147, 1],
    vote_rshares=[147, 1],
    author_rewards=[147, 1],
    author_reputation=[147, 1]
  },
  **user**={ otype=[8394, 1] },
  **comment**={
    otype=[446, 1],
    created=[446, 1],
    last_payout=[446, 1],
    payout=[446, 1],
    net_rshares=[446, 1],
    abs_rshares=[446, 1],
    vote_rshares=[446, 1],
    author_rewards=[446, 1],
    author_reputation=[446, 1]
  },
  (**user, authored, post**)={
    edge_index=[2, 514],
    edge_attr={
      otype=[514, 1],
      time=[514, 1]
    }
  },
  (**user, vote, post**)={
    edge_index=[2, 13572],
    edge_attr={
      otype=[13572, 1],
      time=[13572, 1]
    }
  },
  (**post, reply, comment**)={
    edge_index=[2, 446],
    edge_attr={
      otype=[446, 1],
      time=[446, 1]
    }
  }
)

**Comment by rusty1s** [comment]
The idea here is that you group all your features of a single node type into one feature matrix, e.g., via:
```python
data["post"].x = torch.cat([data["post"].otype, data["post"].created, ...], dim=-1)
```

**Reply by negarmaleki96** [reply]
Thank you so much. This is what I thought, too. So for the edge attributes, as they have 2 attributes each, should I do the same? 

Also, is it okay that my nodes have a different number of features within? (e.g., post node has 9 features, user node has 1 feature, comment node has 9 features, edge attributes have 2 features each.)

**Reply by rusty1s** [reply]
Yes. In a heterogeneous graph, it is okay that features of different type have different number of features.
